### PR TITLE
pull_request action "edited": changes

### PR DESCRIPTION
**Pull request by seregamorph**
# Description 
Webhook with `X-GitHub-Event: pull_request` and `"action": "edited"` has a field `"changes"` that is not present in `GHEventPayload.PullRequest`. I've found three triggers that leads to "edited" action:
- Change PR title:
```json
...
  "changes": {
    "title": {
      "from": "REST-276 - easy-random"
    }
  },
...
```
(full payload - in test data)
- Change PR description:
```json
...
"changes": {
    "body": {
      "from": "**JIRA Ticket URL:**\r\nhttps://jira.devfactory.com/browse/REST-276\r\n\r\n..."
    }
  },
...
```
- Change target branch:
```json
...
  "changes": {
    "base": {
      "ref": {
        "from": "develop"
      },
      "sha": {
        "from": "4b0f3b9fd582b071652ccfccd10bfc8c143cff96"
      }
    }
  },
...
```

So, `"changes"` fields contain old value that was changed, while `pull_request` object has latest values (see UT json payloads). Any other changes of a PR in GitHub UI leads to other types of notifications (e.g. `action` is `assigned` or `ready_for_review` or `labeled`).

### CODE REVIEW

File: src/main/java/org/kohsuke/github/GHPullRequestChanges.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+/**
+ * Wrapper to define changed fields on pull_request action="edited"
+ *
+ * @see GHEventPayload.PullRequest
+ */
+@SuppressFBWarnings("UWF_UNWRITTEN_FIELD")
+public class GHPullRequestChanges {
+
+    private GHCommitPointer base;
```
Comment (seregamorph) [inline src/main/java/org/kohsuke/github/GHPullRequestChanges.java:13]: `base` changed - covered with UT
`title` changed - covered with UT
`body` changed - cover on demand

File: src/test/java/org/kohsuke/github/GHEventPayloadTest.java [hunk 2]
```diff
@@ -214,6 +214,33 @@ public void pull_request() throws Exception {
...
+    public void pull_request_edited_base() throws Exception {
+        GHEventPayload.PullRequest event = GitHub.offline()
+                .parseEventPayload(payload.asReader(), GHEventPayload.PullRequest.class);
+
+        assertThat(event.getAction(), is("edited"));
+        assertThat(event.getChanges().getTitle(), nullValue());
+        assertThat(event.getPullRequest().getTitle(), is("REST-276 - easy-random"));
+        assertThat(event.getChanges().getBase().getRef().getFrom(), is("develop"));
+        assertThat(event.getChanges().getBase().getSha().getFrom(), is("4b0f3b9fd582b071652ccfccd10bfc8c143cff96"));
+        assertThat(event.getPullRequest().getBody(), startsWith("**JIRA Ticket URL:**"));
```
Comment (seregamorph) [inline src/test/java/org/kohsuke/github/GHEventPayloadTest.java:227]: body not changed - hence `"changes"` does not contain it

File: src/test/java/org/kohsuke/github/GHEventPayloadTest.java [hunk 3]
```diff
@@ -214,6 +214,33 @@ public void pull_request() throws Exception {
...
+        assertThat(event.getChanges().getBody(), nullValue());
```
Comment (bitwiseman) [inline src/test/java/org/kohsuke/github/GHEventPayloadTest.java:228]:
```suggestion
        assertThat(event.getChanges().getBody(), nullValue());
        assertThat(event.getChanges().getTitle(), nullValue());
```

File: src/test/java/org/kohsuke/github/GHEventPayloadTest.java [hunk 4]
```diff
@@ -214,6 +214,33 @@ public void pull_request() throws Exception {
...
+    public void pull_request_edited_title() throws Exception {
+        GHEventPayload.PullRequest event = GitHub.offline()
+                .parseEventPayload(payload.asReader(), GHEventPayload.PullRequest.class);
+
+        assertThat(event.getAction(), is("edited"));
+        assertThat(event.getChanges().getTitle().getFrom(), is("REST-276 - easy-random"));
+        assertThat(event.getPullRequest().getTitle(), is("REST-276 - easy-random 4.3.0"));
+        assertThat(event.getChanges().getBase(), nullValue());
+        assertThat(event.getPullRequest().getBody(), startsWith("**JIRA Ticket URL:**"));
+        assertThat(event.getChanges().getBody(), nullValue());
```
Comment (bitwiseman) [inline src/test/java/org/kohsuke/github/GHEventPayloadTest.java:241]:
```suggestion
        assertThat(event.getChanges().getBody(), nullValue());
        assertThat(event.getChanges(). getChanges(), nullValue());
```

**Review by bitwiseman — CHANGES_REQUESTED** [review]
Minor changes to test.

File: src/test/java/org/kohsuke/github/GHEventPayloadTest.java [hunk 5]
```diff
@@ -214,6 +214,35 @@ public void pull_request() throws Exception {
...
+        assertThat(event.getChanges(). getChanges(), nullValue());
```
Comment (bitwiseman) [inline src/test/java/org/kohsuke/github/GHEventPayloadTest.java:243]:
```suggestion
        assertThat(event.getChanges().getTitle(), nullValue());
```

Comment (bitwiseman) [inline src/test/java/org/kohsuke/github/GHEventPayloadTest.java]:
```suggestion
```

File: src/test/java/org/kohsuke/github/GHEventPayloadTest.java [hunk 6]
```diff
@@ -214,6 +214,35 @@ public void pull_request() throws Exception {
...
+        GHEventPayload.PullRequest event = GitHub.offline()
+                .parseEventPayload(payload.asReader(), GHEventPayload.PullRequest.class);
+
+        assertThat(event.getAction(), is("edited"));
+        assertThat(event.getChanges().getTitle().getFrom(), is("REST-276 - easy-random"));
+        assertThat(event.getPullRequest().getTitle(), is("REST-276 - easy-random 4.3.0"));
+        assertThat(event.getChanges().getBase(), nullValue());
+        assertThat(event.getPullRequest().getBody(), startsWith("**JIRA Ticket URL:**"));
+        assertThat(event.getChanges().getBody(), nullValue());
+        assertThat(event.getChanges().getTitle(), nullValue());
```
Comment (bitwiseman) [inline src/test/java/org/kohsuke/github/GHEventPayloadTest.java:243]:
```suggestion
```

**Comment by seregamorph** [comment]
@bitwiseman not sure what changes did you expect. My logic was: for three fields: `title`, `body` and `base` check old value (in `"changes"` field) and current value (in `"pull_request` field). There is a small test enhancement.

**Review by bitwiseman — APPROVED** [review]

**Pull request merged**